src/stitching.rs

```diff
@@ -0,0 +1,430 @@
+        R: std::iter::Extend<Handle<PartialPath>>,
+    {
+        // If the path currently ends at the root node, then we need to look up partial paths whose
+        // symbol stack precondition is compatible with the path.
+        let mut symbol_stack = SymbolStackKey::from_symbol_stack(paths, self, symbol_stack);
+        loop {
+            let key_handle = symbol_stack.back_handle();
+            if let Some(paths) = self.root_paths_by_precondition.get(key_handle) {
+                result.extend(paths.iter().copied());
+            }
+            if symbol_stack.pop_back(self).is_none() {
+                break;
+            }
+        }
+    }
+
+    /// Find all partial paths in the database that start at the given node.  We don't filter the
+    /// results any further than that, since we have to check each partial path for compatibility
+    /// as we try to append it to the current incomplete path anyway, and non-root nodes will
+    /// typically have a small number of outgoing edges.
+    pub fn find_candidate_partial_paths_from_node<R>(
+        &self,
+        start_node: Handle<Node>,
+        result: &mut R,
+    ) where
+        R: std::iter::Extend<Handle<PartialPath>>,
+    {
+        // Return all of the partial paths that start at the requested node.
+        if let Some(paths) = self.paths_by_start_node.get(start_node) {
+            result.extend(paths.iter().copied());
+        }
+    }
+}
+
+impl Index<Handle<PartialPath>> for Database {
+    type Output = PartialPath;
+    #[inline(always)]
+    fn index(&self, handle: Handle<PartialPath>) -> &PartialPath {
+        self.partial_paths.get(handle)
+    }
+}
+
+#[derive(Clone, Copy)]
+struct SymbolStackKey {
+    symbols: List<Handle<Symbol>>,
+}
+
+#[derive(Clone, Eq, Hash, PartialEq)]
+struct SymbolStackCacheKey {
+    head: Handle<Symbol>,
+    tail: SymbolStackKeyHandle,
+}
+
+type SymbolStackKeyCell = ListCell<Handle<Symbol>>;
+type SymbolStackKeyHandle = Handle<SymbolStackKeyCell>;
+
+impl SymbolStackKey {
+    /// Returns an empty symbol stack key.
+    fn empty() -> SymbolStackKey {
+        SymbolStackKey {
+            symbols: List::empty(),
+        }
+    }
+
+    /// Pushes a new symbol onto the back of this symbol stack key.
+    fn push_back(&mut self, db: &mut Database, symbol: Handle<Symbol>) {
+        let cache_key = SymbolStackCacheKey {
+            head: symbol,
+            tail: self.back_handle(),
+        };
+        if let Some(handle) = db.symbol_stack_key_cache.get(&cache_key) {
+            self.symbols = List::from_handle(*handle);
+            return;
+        }
+        self.symbols.push_front(&mut db.symbol_stack_keys, symbol);
+        let handle = self.back_handle();
+        db.symbol_stack_key_cache.insert(cache_key, handle);
+    }
+
+    /// Pops a symbol from the back of this symbol stack key.
+    fn pop_back(&mut self, db: &Database) -> Option<Handle<Symbol>> {
+        self.symbols.pop_front(&db.symbol_stack_keys).copied()
```

Comment:
Should this be `self.symbols.pop_back` instead of `pop_front`?

Reply:
Nope, we're using `List` under the covers, which only has `push/pop_front`.  That means we're storing the content of the key in reverse order.  That ends up not mattering since it's an internal detail; as long as we're consistent throughout the implementation of this type it will all work out.  I'll add a clarifying comment to that effect.